### ISSUE TITLE
cuDNN batchnorm with non-contiguous running mean silently discards updates

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#33867]
```
  if (use_cudnn && eps >= detail::getCUDAHooks().batchnormMinEpsilonCuDNN()) {
    return std::tuple_cat(
             at::cudnn_batch_norm(
               input.contiguous(input.suggest_memory_format()), weight.contiguous(),
               bias.contiguous(),
               running_mean.defined() ? running_mean.contiguous() : running_mean,
               running_var.defined() ? running_var.contiguous() : running_var,
               training, momentum, eps),
             std::make_tuple(1));
  }

```

This looks highly suspect. If running mean/var are not contiguous, then we will allocate a new tensor to make the contiguous, and thereby end up losing updates to the running mean/var entirely.

Discovered while looking into #13402

**Comment by ptrblck** [comment]
Thanks for raising this issue.
After discussing it with @jjsjann123, we think checking the running stats for `is_contiguous()` and making them contiguous could be added before the `cudnn_batch_norm` call, as they are passed as references.

Let me know, if that would work or if we're missing something, and we can create the PR.

**Comment by ezyang** [comment]
@ptrblck I don't think that works, because if there is ever a `contiguous()` call on a non-contiguous tensor you'll end up with a new tensor that isn't contiguous, and so updates won't be reflected in the original. In cases like this, you need to explicitly `copy_` the results back to the original tensor.

BTW, I doubt anyone is actually using noncontiguous running mean, so I don't think this is terribly high priority.

**Comment by jjsjann123** [comment]
Good catch! I got confused at the python binding part, updating `running_mean = running_mean.contiguous()` would not update the buffer in the layer to the new one. And I don't know if there's any way to do that. I stand corrected and we do need to update the original tensor.

`copy_` would work. We could add that in the `cudnn_batch_norm` if we wanted the support. But that means unnecessary copy for everything FW call without user even knowing. We can discuss whether we want to support this.

For the time being, I think we should simply removing the contiguous call in `running_mean.defined() ? running_mean.contiguous() : running_mean` and forward `running_mean` directly. cudnn will throw an error as running_mean is not contiguous. So at least it's not a silent wrong result. But that means existing code hitting non-contiguous running_mean/var would be broken. So maybe we do need to have `copy_` in the code as we fix this.

**Comment by ezyang** [comment]
Yeah, the point is you'd only copy_ if you noticed it was non-contiguous; so no performance hit in the fast case.

**Comment by VitalyFedyunin** [comment]
Seems to me miopen issue as well:
https://github.com/pytorch/pytorch/blob/acbca57d18336405fc66bb63f28bbe8eb47d937c/aten/src/ATen/native/Normalization.cpp#L453

As we are discarding updated var and mean here:
https://github.com/pytorch/pytorch/blob/acbca57d18336405fc66bb63f28bbe8eb47d937c/aten/src/ATen/native/Normalization.cpp#L496
